schema/schema.go

```diff
@@ -1,15 +1,167 @@
 package schema
 
-import "github.com/twitchscience/scoop_protocol/scoop_protocol"
+import "errors"
 
-type UpdateSchemaRequest struct {
-	EventName string `json:"-"`
-	Columns   []scoop_protocol.ColumnDefinition
+type Migrator interface {
+	Validate() (*Event, error)
 }
 
-func (u *UpdateSchemaRequest) ConvertToRedshiftUpdate() *scoop_protocol.Config {
-	return &scoop_protocol.Config{
-		EventName: u.EventName,
-		Columns:   u.Columns,
+type Event struct {
+	Name            string
+	Version         int
+	ColumnSchema    ColumnSchema
+	ParentMigration Migration
+}
+
+type ColumnSchema struct {
+	TableOption TableOption
+	Columns     []ColumnDefinition
+}
+
+type ColumnDefinition struct {
+	InboundName           string
+	OutboundName          string
+	Transformer           string
+	ColumnCreationOptions string
+}
+
+type Migration struct {
+	TableOperation   string
+	Name             string
+	ColumnOperations []ColumnOperation
+	TableOption      TableOption
+}
+
+type ColumnOperation struct {
+	Operation           string
+	InboundName         string
+	OutboundName        string
+	NewColumnDefinition ColumnDefinition
+}
+
+type TableOption struct {
+	DistKey []string
+	SortKey []string
+}
+
+func (s *ColumnSchema) IsEmpty() bool {
+	if len(s.Columns) == 0 {
+		return true
+	}
+	return false
+}
+
+func (to *TableOption) IsEmpty() bool {
+	if len(to.DistKey) == 0 && len(to.SortKey) == 0 {
+		return true
+	}
+	return false
+}
+
+func (e *Event) AddColumn(ColumnOperation ColumnOperation) error {
```

Comment:
The uppercase name here shadows the type; I think you want the value to be columnOperation. You repeat this pattern elsewhere.